File: apps/ios/Sources/TePlannerKit/Services/HubActionsStore.swift
import Foundation
import Combine

/// Owns the user's Hub Quick Actions: the *pool* of defined actions
/// plus which ones occupy the 8 on-screen *slots*. Persists to the
/// backend `user_settings` bag under keys `hub.actions` and
/// `hub.slots`, so the same definitions appear after reinstall and
/// (once Android lands the same client) on every other device.
///
/// Lifecycle:
///   - Hub creates one store on appear, calls `load()` which fetches
///     `/user/settings`.
///   - If the bag has no `hub.*` keys (fresh user), the store seeds
///     four default system actions in the first row and persists.
///   - Every mutate (`create`, `update`, `delete`, `assignSlot`,
///     `moveSlot`) writes the affected key back via PUT.
///   - `run(_:vehicleId:)` walks the steps, dispatches via
///     `APIServiceProtocol.invokeCapability`, optionally waits the
///     `delayMsAfter` between steps.
///
/// Concurrency: marked @MainActor because Compose / SwiftUI reads
/// @Published, and the load/save coroutines do their own awaits so
/// nothing blocks the main thread despite the actor isolation.
@MainActor
public final class HubActionsStore: ObservableObject {

    // MARK: - Published state

    @Published public private(set) var actions: [HubAction] = []
    @Published public private(set) var slots: HubSlots = HubSlots()
    @Published public private(set) var isLoading: Bool = false
    @Published public private(set) var lastError: String?
    /// Most recent action that was dispatched. Hub uses this to
    /// trigger the existing CommandStatusBanner poll → user gets
    /// "切换哨兵中… → 已确认" feedback identical to chip taps.
    @Published public private(set) var lastDispatchedActionId: String?

    // MARK: - Settings bag keys

    public static let actionsKey: String = "hub.actions"
    public static let slotsKey: String = "hub.slots"

    // MARK: - Dependencies

    private let apiService: APIServiceProtocol

    public init(apiService: APIServiceProtocol) {
        self.apiService = apiService
    }

    // MARK: - Lookup

    public func action(id: String) -> HubAction? {
        actions.first { $0.id == id }
    }

    /// The HubAction occupying a given on-screen slot (0..<8), or
    /// nil if the slot is empty or refers to an action that has
    /// been deleted from the pool.
    public func slotAction(at index: Int) -> HubAction? {
        guard slots.slots.indices.contains(index) else { return nil }
        guard let id = slots.slots[index] else { return nil }
        return action(id: id)
    }

    // MARK: - Load / save

    public func load() async {
        isLoading = true
        defer { isLoading = false }

        switch await apiService.getUserSettings() {
        case .failure(let err):
            lastError = err.localizedDescription
        case .success(let bag):
            decode(bag: bag)
            // Fresh user: seed defaults exactly once. Detect via
            // "neither key present" so we don't re-seed if the user
            // intentionally emptied all slots.
            if bag[Self.actionsKey] == nil, bag[Self.slotsKey] == nil {
                seedDefaults()
                await persistAll()
            }
        }
    }

    private func decode(bag: [String: JSONValue]) {
        if case let .array(arr)? = bag[Self.actionsKey] {
            let raw = JSONValue.array(arr)
            if let data = try? JSONEncoder().encode(raw),
               let decoded = try? JSONDecoder().decode([HubAction].self, from: data) {
                actions = decoded
            }
        }
        if case let .object(obj)? = bag[Self.slotsKey] {
            let raw = JSONValue.object(obj)
            if let data = try? JSONEncoder().encode(raw),
               let decoded = try? JSONDecoder().decode(HubSlots.self, from: data) {
                slots = decoded
            }
        }
    }

    private func persistAll() async {
        let payload: [String: JSONValue] = [
            Self.actionsKey: encodeActions(),
            Self.slotsKey: encodeSlots(),
        ]
        _ = await apiService.putUserSettings(payload, replaceAll: false)
    }

    private func persistActions() async {
        _ = await apiService.putUserSettings(
            [Self.actionsKey: encodeActions()], replaceAll: false
        )
    }

    private func persistSlots() async {
        _ = await apiService.putUserSettings(
            [Self.slotsKey: encodeSlots()], replaceAll: false
        )
    }

    // MARK: - JSON helpers (round-trip through JSONEncoder so the
    // Codable conformances on HubAction / HubSlots stay the source
    // of truth — no parallel hand-rolled encoding to drift from).

    private func encodeActions() -> JSONValue {
        guard let data = try? JSONEncoder().encode(actions),
              let val = try? JSONDecoder().decode(JSONValue.self, from: data) else {
            return .array([])
        }
        return val
    }

    private func encodeSlots() -> JSONValue {
        guard let data = try? JSONEncoder().encode(slots),
              let val = try? JSONDecoder().decode(JSONValue.self, from: data) else {
            return .object([:])
        }
        return val
    }

    // MARK: - Mutations

    /// Insert a new action and (optionally) drop it in the first
    /// empty slot. Returns the new action id.
    @discardableResult
    public func create(
        name: String,
        icon: String,
        tint: HubActionTint,
        steps: [HubActionStep],
        confirmRequired: Bool,
        assignToFirstEmpty: Bool = true
    ) async -> String {
        let action = HubAction(
            name: name, icon: icon, tint: tint,
            steps: steps, confirmRequired: confirmRequired,
        )
        actions.append(action)
        if assignToFirstEmpty,
           let idx = slots.slots.firstIndex(of: nil) {
            slots.slots[idx] = action.id
        }
        await persistAll()
        return action.id
    }

    /// Replace fields on an existing action. System actions can have
    /// name / icon / tint / confirmRequired / steps changed but not
    /// `isSystem` itself (caller can't promote a custom to system).
    public func update(
        id: String,
        name: String,
        icon: String,
        tint: HubActionTint,
        steps: [HubActionStep],
        confirmRequired: Bool
    ) async {
        guard let idx = actions.firstIndex(where: { $0.id == id }) else { return }
        let old = actions[idx]
        actions[idx] = HubAction(
            id: old.id,
            name: name,
            icon: icon,
            tint: tint,
            steps: steps,
            confirmRequired: confirmRequired,
            isSystem: old.isSystem,
            createdAt: old.createdAt,
        )
        await persistActions()
    }

    /// Remove a custom action. No-op for system actions (UI hides
    /// the delete button on those, but defend in depth here too).
    public func delete(id: String) async {
        guard let target = action(id: id), !target.isSystem else { return }
        actions.removeAll { $0.id == id }
        // Any slot pointing at this action becomes empty.
        for i in slots.slots.indices where slots.slots[i] == id {
            slots.slots[i] = nil
        }
        await persistAll()
    }

    /// Drop an action into a specific slot. Pass `nil` actionId to
    /// clear. If `actionId` was already in another slot, that other
    /// slot is cleared (an action can only occupy one slot at a time).
    public func assignSlot(index: Int, actionId: String?) async {
        guard slots.slots.indices.contains(index) else { return }
        if let actionId {
            for i in slots.slots.indices where slots.slots[i] == actionId && i != index {
                slots.slots[i] = nil
            }
        }
        slots.slots[index] = actionId
        await persistSlots()
    }

    /// Drag-and-drop reorder. Swaps the two slot positions. Either
    /// can be empty; an empty source means we're filling an empty
    /// destination from elsewhere.
    public func swapSlots(_ a: Int, _ b: Int) async {
        guard slots.slots.indices.contains(a),
              slots.slots.indices.contains(b),
              a != b else { return }
        slots.slots.swapAt(a, b)
        await persistSlots()
    }

    // MARK: - Dispatch

    public enum RunError: Error, Equatable {
        case noVehicle
        case stepFailed(stepIndex: Int, message: String)
        case unknownAction
    }

    /// Execute an action against the given vehicle. Walks steps in
    /// order; aborts on the first failure (returns `.stepFailed(idx)`).
    /// `confirmRequired` is the *caller's* job — the UI shows the
    /// confirm alert before invoking this. We don't gate here so
    /// callers can suppress it (e.g. a future Siri shortcut).
    @discardableResult
    public func run(
        actionId: String,
        vehicleId: String
    ) async -> Result<Void, RunError> {
        guard let action = action(id: actionId) else {
            return .failure(.unknownAction)
        }
        lastDispatchedActionId = actionId
        for (idx, step) in action.steps.enumerated() {
            let result = await apiService.invokeCapability(
                vehicleId: vehicleId,
                capability: step.capability,
                params: step.params,
            )
            switch result {
            case .failure(let err):
                return .failure(.stepFailed(
                    stepIndex: idx,
                    message: err.localizedDescription,
                ))
            case .success:
                // Inter-step delay. Tested with a fake clock injected
                // by callers; default uses Task.sleep which is fine in
                // production.
                if let ms = step.delayMsAfter, ms > 0, idx < action.steps.count - 1 {
                    try? await Task.sleep(nanoseconds: UInt64(ms) * 1_000_000)
                }
            }
        }
        return .success(())
    }

    // MARK: - Default seeding

    /// Four common single-step actions, dropped in the top row on
    /// first launch. The user is free to delete / rearrange / replace
    /// any of them; we don't re-seed once user_settings has *any*
    /// hub.* keys.
    private func seedDefaults() {
        let presets: [HubAction] = [
            HubAction(
                id: "system_lock",
                name: "锁车",
                icon: "lock.fill",
                tint: .blue,
                steps: [HubActionStep(capability: "tesla.security.door_lock")],
                confirmRequired: false,
                isSystem: true,
            ),
            HubAction(
                id: "system_unlock",
                name: "解锁",
                icon: "lock.open.fill",
                tint: .red,
                steps: [HubActionStep(capability: "tesla.security.door_unlock")],
                confirmRequired: true,
                isSystem: true,
            ),
            HubAction(
                id: "system_preheat",
                name: "预热",
                icon: "thermometer.medium",
                tint: .orange,
                steps: [HubActionStep(capability: "tesla.climate.preheat")],
                confirmRequired: false,
                isSystem: true,
            ),
            HubAction(
                id: "system_trunk",
                name: "后备箱",
                icon: "suitcase.fill",
                tint: .blue,
                steps: [HubActionStep(capability: "tesla.security.actuate_trunk")],
                confirmRequired: true,
                isSystem: true,
            ),
        ]
        actions = presets
        // Fill the first 4 slots; row 2 stays empty for the user to fill.
        slots = HubSlots(slots: presets.map { $0.id } + Array(repeating: nil, count: 4))
    }
}

File: apps/android/app/src/main/java/cloud/teplanner/android/hub/quickactions/HubActionsStore.kt
package cloud.teplanner.android.hub.quickactions

import android.util.Log
import cloud.teplanner.android.core.network.UserApi
import cloud.teplanner.android.core.network.UserSettingsRequest
import dagger.hilt.android.scopes.ViewModelScoped
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import java.util.UUID
import javax.inject.Inject

/**
 * Kotlin port of iOS `HubActionsStore`. Single source of truth for
 * the user's Quick Actions library (`actions`) and on-Hub slot
 * assignments (`slots`). Persists to backend `/user/settings` under
 * keys `hub.actions` + `hub.slots` — same keys iOS uses, so the
 * two clients can share state when the same user signs in on both
 * platforms.
 *
 * Public mutators (create / update / delete / assignSlot / swapSlots /
 * resetToDefaults / importAction) update the StateFlow synchronously
 * then fire off a persistAll() to push to the backend. Loaders are
 * forgiving — unknown keys / malformed JSON fall back to the seeded
 * defaults rather than crashing the Hub.
 */
@ViewModelScoped
class HubActionsStore @Inject constructor(
    private val userApi: UserApi,
) {

    private val json: Json = Json {
        ignoreUnknownKeys = true
        encodeDefaults = true
    }

    private val _actions = MutableStateFlow<List<HubAction>>(emptyList())
    val actions: StateFlow<List<HubAction>> = _actions.asStateFlow()

    private val _slots = MutableStateFlow(HubSlots())
    val slots: StateFlow<HubSlots> = _slots.asStateFlow()

    fun action(id: String): HubAction? = _actions.value.firstOrNull { it.id == id }

    fun slotAction(index: Int): HubAction? {
        val slotId = _slots.value.slots.getOrNull(index) ?: return null
        return action(slotId)
    }

    suspend fun load() {
        val resp = runCatching { userApi.getUserSettings() }.getOrNull()
        val bag = resp?.settings ?: emptyMap()

        if (HUB_ACTIONS_KEY !in bag && HUB_SLOTS_KEY !in bag) {
            seedDefaults()
            persistAll()
            return
        }

        val decodedActions = decodeActions(bag[HUB_ACTIONS_KEY])
        val decodedSlots = decodeSlots(bag[HUB_SLOTS_KEY])
        _actions.value = decodedActions
        _slots.value = decodedSlots
    }

    suspend fun create(
        name: String,
        icon: String,
        tint: HubActionTint,
        steps: List<HubActionStep>,
        confirmRequired: Boolean,
        assignToFirstEmpty: Boolean = true,
    ): String {
        val action = HubAction(
            name = name,
            icon = icon,
            tint = tint,
            steps = steps,
            confirmRequired = confirmRequired,
        )
        _actions.update { it + action }
        if (assignToFirstEmpty) {
            val firstEmpty = _slots.value.slots.indexOfFirst { it == null }
            if (firstEmpty >= 0) {
                _slots.value = _slots.value.withSlot(firstEmpty, action.id)
            }
        }
        persistAll()
        return action.id
    }

    /**
     * Insert an action received via share code. The caller already
     * generated a fresh id (HubAction's default UUID); we only drop
     * it into the library + persist. Imported actions are NEVER
     * auto-assigned to a slot — the user explicitly chooses.
     *
     * Name collision handling: if any existing action (system or
     * custom) shares this name, append "  副本" / " 副本 2" / ... so
     * the picker row stays distinguishable. Mirrors iOS commit
     * cfd6556.
     */
    suspend fun importAction(action: HubAction) {
        val uniqueName = disambiguateImportedName(action.name)
        val idCollision = _actions.value.any { it.id == action.id }
        val importing = action.copy(
            id = if (idCollision) UUID.randomUUID().toString() else action.id,
            name = uniqueName,
            isSystem = false, // imports are always user-owned
        )
        _actions.update { it + importing }
        persistAll()
    }

    fun disambiguateImportedName(base: String): String {
        val taken = _actions.value.map { it.name }.toSet()
        if (base !in taken) return base
        var candidate = "$base 副本"
        var counter = 2
        while (candidate in taken) {
            candidate = "$base 副本 $counter"
            counter += 1
        }
        return candidate
    }

    suspend fun update(
        id: String,
        name: String,
        icon: String,
        tint: HubActionTint,
        steps: List<HubActionStep>,
        confirmRequired: Boolean,
    ) {
        _actions.update { list ->
            list.map { a ->
                if (a.id == id) a.copy(
                    name = name,
                    icon = icon,
                    tint = tint,
                    steps = steps,
                    confirmRequired = confirmRequired,
                ) else a
            }
        }
        persistAll()
    }

    suspend fun delete(id: String) {
        val target = action(id) ?: return
        if (target.isSystem) return
        _actions.update { it.filter { a -> a.id != id } }
        // Clear any slot referencing this action.
        _slots.update { current ->
            val cleared = current.slots.map { if (it == id) null else it }
            HubSlots(cleared)
        }
        persistAll()
    }

    suspend fun assignSlot(index: Int, actionId: String?) {
        if (index !in 0 until HubSlots.COUNT) return
        _slots.update { current ->
            var next = current
            if (actionId != null) {
                // An action can only occupy one slot at a time —
                // clear any other slot already holding it.
                val cleared = current.slots.mapIndexed { i, id ->
                    if (i != index && id == actionId) null else id
                }
                next = HubSlots(cleared)
            }
            next.withSlot(index, actionId)
        }
        persistSlots()
    }

    suspend fun swapSlots(a: Int, b: Int) {
        if (a !in 0 until HubSlots.COUNT || b !in 0 until HubSlots.COUNT || a == b) {
            return
        }
        _slots.update { it.swap(a, b) }
        persistSlots()
    }

    suspend fun resetToDefaults() {
        _actions.value = emptyList()
        _slots.value = HubSlots()
        seedDefaults()
        persistAll()
    }

    // MARK: - Default seeding (mirrors iOS HubActionsStore.seedDefaults)

    private fun seedDefaults() {
        val presets = listOf(
            HubAction(
                id = "system_lock",
                name = "锁车",
                icon = "lock.fill",
                tint = HubActionTint.BLUE,
                steps = listOf(HubActionStep(capability = "tesla.security.door_lock")),
                confirmRequired = false,
                isSystem = true,
            ),
            HubAction(
                id = "system_unlock",
                name = "解锁",
                icon = "lock.open.fill",
                tint = HubActionTint.RED,
                steps = listOf(HubActionStep(capability = "tesla.security.door_unlock")),
                confirmRequired = true,
                isSystem = true,
            ),
            HubAction(
                id = "system_preheat",
                name = "预热",
                icon = "thermometer.medium",
                tint = HubActionTint.ORANGE,
                steps = listOf(HubActionStep(capability = "tesla.climate.preheat")),
                confirmRequired = true,
                isSystem = true,
            ),
            HubAction(
                id = "system_trunk",
                name = "后备箱",
                icon = "suitcase.fill",
                tint = HubActionTint.BLUE,
                steps = listOf(HubActionStep(capability = "tesla.security.actuate_trunk")),
                confirmRequired = true,
                isSystem = true,
            ),
        )
        _actions.value = presets
        _slots.value = HubSlots(
            slots = listOf(
                presets[0].id, presets[1].id, presets[2].id, presets[3].id,
                null, null, null, null,
            )
        )
    }

    // MARK: - Persistence helpers

    private suspend fun persistAll() {
        val bag = mapOf(
            HUB_ACTIONS_KEY to encodeActions(),
            HUB_SLOTS_KEY to encodeSlots(),
        )
        runCatching {
            userApi.putUserSettings(UserSettingsRequest(settings = bag, replaceAll = false))
        }.onFailure { Log.w(TAG, "persistAll failed: ${it.message}") }
    }

    private suspend fun persistSlots() {
        val bag = mapOf(HUB_SLOTS_KEY to encodeSlots())
        runCatching {
            userApi.putUserSettings(UserSettingsRequest(settings = bag, replaceAll = false))
        }.onFailure { Log.w(TAG, "persistSlots failed: ${it.message}") }
    }

    private fun encodeActions(): JsonElement =
        json.encodeToJsonElement(kotlinx.serialization.serializer(), _actions.value)

    private fun encodeSlots(): JsonElement =
        json.encodeToJsonElement(kotlinx.serialization.serializer(), _slots.value)

    private fun <T> Json.encodeToJsonElement(
        serializer: kotlinx.serialization.KSerializer<T>,
        value: T,
    ): JsonElement {
        val text = this.encodeToString(serializer, value)
        return this.parseToJsonElement(text)
    }

    private fun decodeActions(raw: JsonElement?): List<HubAction> {
        if (raw == null) return emptyList()
        return runCatching {
            json.decodeFromJsonElement<List<HubAction>>(
                kotlinx.serialization.serializer(),
                raw,
            )
        }.getOrElse {
            Log.w(TAG, "decodeActions failed: ${it.message}")
            emptyList()
        }
    }

    private fun decodeSlots(raw: JsonElement?): HubSlots {
        if (raw == null) return HubSlots()
        return runCatching {
            // iOS encodes HubSlots as { "slots": [...] }. Tolerate
            // raw arrays too (a previous Android build might have
            // emitted that shape) by checking the wire shape first.
            val obj = raw.jsonObject
            val arr = obj["slots"]?.jsonArray
                ?: return@runCatching HubSlots.normalized(emptyList())
            val ids = arr.map { el ->
                val prim = runCatching { el.jsonPrimitive }.getOrNull()
                if (prim == null || prim.isString.not()) null else prim.content
            }
            HubSlots.normalized(ids)
        }.getOrElse {
            Log.w(TAG, "decodeSlots failed: ${it.message}")
            HubSlots()
        }
    }

    private fun <T> Json.decodeFromJsonElement(
        serializer: kotlinx.serialization.KSerializer<T>,
        element: JsonElement,
    ): T {
        val text = element.toString()
        return this.decodeFromString(serializer, text)
    }

    companion object {
        private const val TAG = "HubActionsStore"
        const val HUB_ACTIONS_KEY = "hub.actions"
        const val HUB_SLOTS_KEY = "hub.slots"
    }
}
